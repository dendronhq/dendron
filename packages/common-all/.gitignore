coverage/
node_modules/
npm-debug.log*
.nyc_output
lib
lib_test
