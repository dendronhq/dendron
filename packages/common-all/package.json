{
  "name": "@dendronhq/common-all",
  "version": "0.124.0",
  "description": "common-all",
  "license": "Apache 2.0",
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/dendronhq/dendron.git",
    "directory": "packages/common-all"
  },
  "author": {
    "name": "Kevin Lin",
    "email": "kevin@thence.io",
    "url": "https://kevinslin.com"
  },
  "keywords": [
    "common"
  ],
  "files": [
    "lib"
  ],
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "clean": "rimraf lib && rimraf coverage",
    "format": "echo nop",
    "lint": "echo stub",
    "prepublishOnly": "yarn build",
    "prebuild": "yarn clean && yarn format && yarn lint && echo Using TypeScript && tsc --version",
    "build": "yarn compile",
    "buildCI": "yarn compile",
    "compile": "tsc -p tsconfig.build.json ",
    "watch": "yarn compile --watch"
  },
  "dependencies": {
    "@swc/helpers": "^0.3.2",
    "axios": "^0.21.4",
    "dropbox": "^4.0.30",
    "fast-levenshtein": "^3.0.0",
    "fuse.js": "^6.4.6",
    "github-slugger": "^1.3.0",
    "gray-matter": "^4.0.2",
    "http-status-codes": "^2.1.4",
    "js-yaml": "^4.1.0",
    "lodash": "^4.17.15",
    "luxon": "^1.25.0",
    "minimatch": "5.1.0",
    "nanoid": "^3.1.23",
    "neverthrow": "^5.0.0",
    "normalize-path": "^3.0.0",
    "qs": "^6.10.1",
    "semver": "^7.3.2",
    "spark-md5": "^3.0.2",
    "title": "^3.4.4",
    "vscode-uri": "^3.0.3",
    "zod": "3.19.1",
    "zod-validation-error": "^0.1.1"
  },
  "devDependencies": {
    "@types/fast-levenshtein": "^0.0.2",
    "@types/github-slugger": "^1.3.0",
    "@types/js-yaml": "^4.0.5",
    "@types/lodash": "^4.14.152",
    "@types/luxon": "^1.25.0",
    "@types/minimatch": "^3.0.5",
    "@types/nanoid-dictionary": "^4.2.0",
    "@types/node": "13.11.0",
    "@types/normalize-path": "^3.0.0",
    "@types/semver": "^7.3.4",
    "@types/title": "^3.4.1",
    "@types/unist": "^2.0.3",
    "coveralls": "^3.0.2",
    "rimraf": "^2.6.2",
    "tslint": "^5.11.0",
    "tslint-config-prettier": "^1.15.0"
  },
  "engines": {
    "node": ">=12.0.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
