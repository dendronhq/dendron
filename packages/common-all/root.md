---
id: e335a6a8-dca4-4ca5-9842-c6955b953ec1
title: Root
desc: ""
updated: 1609029884026
created: 1609029884026
---
