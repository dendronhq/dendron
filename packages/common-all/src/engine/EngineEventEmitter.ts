import { Disposable, Event, NoteChangeEntry } from "../types";

/**
 * Interface providing events signaling changes that have been made to engine
 * state
 */
export interface EngineEventEmitter extends Disposable {
  /**
   * Event that fires upon the changing of note state in the engine.
   */
  get onEngineNoteStateChanged(): Event<NoteChangeEntry[]>;
}
