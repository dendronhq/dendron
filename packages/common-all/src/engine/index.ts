export * from "./EngineEventEmitter";
export * from "./EngineV3Base";
