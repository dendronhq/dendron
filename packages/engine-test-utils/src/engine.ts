import {
  CleanDendronSiteConfig,
  CONSTANTS, DendronConfig, DEngineClient,
  DVault,
  DWorkspace,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings
} from "@dendronhq/common-all";
import {
  getDurationMilliseconds,
  tmpDir,
  vault2Path
} from "@dendronhq/common-server";
import {
  GenTestResults,
  PostSetupHookFunction,
  PreSetupHookFunction,
  RunEngineTestFunctionOpts,
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
  SetupTestFunctionV4,
  sinon,
  TestResult
} from "@dendronhq/common-test-utils";
import { LaunchEngineServerCommand } from "@dendronhq/dendron-cli";
import {
  createEngine as engineServerCreateEngine,
  DConfig,
  WorkspaceService
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import os from "os";
import path from "path";
import { SinonStub } from "sinon";
import { ENGINE_HOOKS } from "./presets";
import { GitTestUtils } from "./utils";

export type TestSetupWorkspaceOpts = {
  /**
   * Vaults to initialize engine with
   * Defaults to following if not set
   * [
   *    { fsPath: "vault1" },
   *    { fsPath: "vault2" },
   *    { fsPath: "vault3", name: "vaultThree" },
   *  ]
   */
  vaults?: DVault[];
  /**
   * Modify dendron config before initialization
   */
  modConfigCb?: (config: DendronConfig) => DendronConfig;
};

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<{engine: DEngineClient, port?: number}>;

/**
 * Create an {@link DendronEngine}
 */
export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return {engine: engineServerCreateEngine(opts) as DEngineClient, port: undefined};
}

export { DEngineClient, DVault, WorkspaceOpts };

/**
 * Create a server
 * @param opts
 * @returns
 */
// @ts-ignore
export async function createServer(opts: WorkspaceOpts) {
  return await new LaunchEngineServerCommand().enrichArgs({
    wsRoot: opts.wsRoot,
  });
}

/**
 * Create an {@link DendronEngineClient}
 */
export async function createEngineFromServer(
  opts: WorkspaceOpts
): Promise<any> {
  const { engine, port } = await createServer(opts);
  await engine.init();
  return {engine, port};
}

export function createSiteConfig(
  opts: Partial<CleanDendronSiteConfig> &
    Required<Pick<CleanDendronSiteConfig, "siteRootDir" | "siteHierarchies">>
): CleanDendronSiteConfig {
  let copts = {
    siteNotesDir: "docs",
    siteUrl: "https://localhost:8080",
    ...opts,
  };
  return {
    ...copts,
    siteIndex: DConfig.getSiteIndex(copts),
  };
}

/**
 *
 * @param opts.asRemote: add git repo
 * @param opts.wsRoot: override given wsRoot
 * @returns
 */
export async function setupWS(opts: {
  vaults: DVault[];
  workspaces?: DWorkspace[];
  asRemote?: boolean;
  wsRoot?: string;
}) {
  const wsRoot = opts.wsRoot || tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  ws.createConfig();
  const config = ws.config;
  let vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault, config, updateConfig: false });
      return vault;
    })
  );
  config.vaults = _.sortBy(config.vaults, "fsPath");
  if (config.site.duplicateNoteBehavior) {
    config.site.duplicateNoteBehavior.payload = (
      config.site.duplicateNoteBehavior.payload as string[]
    ).sort();
  }
  ws.setConfig(config);
  if (opts.workspaces) {
    const vaultsFromWs = await _.reduce(
      opts.workspaces,
      async (resp, ent) => {
        await resp;
        await WorkspaceService.createWorkspace({
          wsRoot: path.join(wsRoot, ent.name),
          vaults: ent.vaults,
        });
        return ws.addWorkspace({ workspace: ent });
      },
      Promise.resolve({ vaults: [] } as { vaults: DVault[] })
    );
    vaults = vaults.concat(vaultsFromWs.vaults);
  }
  if (opts.asRemote) {
    await GitTestUtils.createRepoWithReadme(wsRoot);
  }
  return { wsRoot, vaults };
}

export type RunEngineTestV5Opts = {
  preSetupHook?: SetupHookFunction;
  createEngine?: AsyncCreateEngineFunction;
  extra?: any;
  expect: any;
  workspaces?: DWorkspace[];
  setupOnly?: boolean;
  initGit?: boolean;
  initHooks?: boolean;
  addVSWorkspace?: boolean;
  /**
   * Path to preset wsRoot
   */
  wsRoot?: string;
} & TestSetupWorkspaceOpts;

export type RunEngineTestFunctionV5<T = any> = (
  opts: RunEngineTestFunctionOpts & { extra?: any; engineInitDuration: number, port?: number }
) => Promise<TestResult[] | void | T>;

export class TestPresetEntryV5 {
  public preSetupHook: PreSetupHookFunction;
  public postSetupHook: PostSetupHookFunction;
  public testFunc: RunEngineTestFunctionV4;
  public extraOpts: any;
  public setupTest?: SetupTestFunctionV4;
  public genTestResults?: GenTestResults;
  public vaults: DVault[];
  public workspaces: DWorkspace[];

  constructor(
    func: RunEngineTestFunctionV5,
    opts?: {
      preSetupHook?: PreSetupHookFunction;
      postSetupHook?: PostSetupHookFunction;
      extraOpts?: any;
      setupTest?: SetupTestFunctionV4;
      genTestResults?: GenTestResults;
      vaults?: DVault[];
    }
  ) {
    let {
      preSetupHook,
      postSetupHook,
      extraOpts,
      setupTest,
      genTestResults,
      workspaces,
    } = _.defaults(opts, {
      workspaces: [],
    });
    this.preSetupHook = preSetupHook ? preSetupHook : async () => {};
    this.postSetupHook = postSetupHook ? postSetupHook : async () => {};
    this.testFunc = _.bind(func, this);
    this.extraOpts = extraOpts;
    this.setupTest = setupTest;
    this.genTestResults = _.bind(
      genTestResults ? genTestResults : async () => [],
      this
    );
    this.workspaces = workspaces;
    this.vaults = opts?.vaults || [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      {
        name: "vaultThree",
        fsPath: "vault3",
      },
    ];
  }
}

/**
 *
 * To create empty workspace, initilizae with `vaults = []`
 * @param func
 * @param opts.vaults: By default, initiate 3 vaults {vault1, vault2, (vault3, "vaultThree")}
 * @param opts.preSetupHook: By default, initiate empty
 * @param opts.wsRoot: Override the randomly generated test directory for the wsRoot
 * @returns
 */
export async function runEngineTestV5(
  func: RunEngineTestFunctionV5,
  opts: RunEngineTestV5Opts
): Promise<any> {
  const {
    preSetupHook,
    extra,
    vaults: vaultsInit,
    createEngine,
    initGit,
    workspaces,
    addVSWorkspace,
  } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    // third vault has diff name
    vaults: [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ],
    addVSWorkspace: false,
  });
  try {
    // make sure tests don't overwrite local homedir contents
    TestEngineUtils.mockHomeDir();
    const { wsRoot, vaults } = await setupWS({
      vaults: vaultsInit,
      workspaces,
      wsRoot: opts.wsRoot,
    });
    if ((opts.initHooks, vaults)) {
      fs.ensureDirSync(path.join(wsRoot, CONSTANTS.DENDRON_HOOKS_BASE));
    }
    await preSetupHook({ wsRoot, vaults });
    const resp = await createEngine({ wsRoot, vaults });
    const engine = resp.engine;
    const start = process.hrtime();
    const initResp = await engine.init();
    if (addVSWorkspace) {
      fs.writeJSONSync(
        path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME),
        {
          folders: vaults.map((ent) => ({
            path: ent.fsPath,
            name: ent.name,
          })) as WorkspaceFolderRaw[],
          settings: {},
          extensions: {},
        } as WorkspaceSettings,
        { spaces: 4 }
      );
    }
    const engineInitDuration = getDurationMilliseconds(start);
    const testOpts = {
      wsRoot,
      vaults,
      engine,
      initResp,
      port: resp.port,
      extra,
      config: engine,
      engineInitDuration,
    };
    if (initGit) {
      await GitTestUtils.createRepoForWorkspace(wsRoot);
      await Promise.all(
        vaults.map((vault) => {
          return GitTestUtils.createRepoWithReadme(
            vault2Path({ vault, wsRoot })
          );
        })
      );
    }
    if (opts.setupOnly) {
      return testOpts;
    }
    const results = (await func(testOpts)) || [];
    await runJestHarnessV2(results, expect);
    return { opts: testOpts, resp: undefined, wsRoot };
  } finally {
    // restore sinon so other tests can keep running
    sinon.restore();
  }
}

export function testWithEngine(
  prompt: string,
  func: RunEngineTestFunctionV4,
  opts?: Omit<RunEngineTestV5Opts, "expect"> & { only?: boolean }
) {
  if (opts?.only) {
    return test.only(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  } else {
    return test(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  }
}

export class TestEngineUtils {
  static mockHomeDir(dir?: string): SinonStub {
    if (_.isUndefined(dir)) dir = tmpDir().name;
    return sinon.stub(os, "homedir").returns(dir);
  }

  static vault1(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault1" })!;
  }

  static vault2(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault2" })!;
  }

  static vault3(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault3" })!;
  }
}
