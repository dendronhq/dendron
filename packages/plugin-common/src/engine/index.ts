export * from "./IReducedEngineApiService";
export * from "./DendronEngineV3Web";
export * from "./store";
