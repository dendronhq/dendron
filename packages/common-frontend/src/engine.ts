import {
  IntermediateDendronConfig,
  DEngineClient,
  DVault,
  NotePropsDict,
  SchemaModuleDict,
} from "@dendronhq/common-all";
import { EngineState } from "./features/engine/slice";

// @ts-ignore
export class WebEngine implements DEngineClient {
  public notes: NotePropsDict;
  public wsRoot: string;
  public schemas: SchemaModuleDict;
  public configRoot: string;
  public vaults: DVault[];
  // public links: DLink[];
  // public fuseEngine: FuseEngine;
  // public api: DendronAPI;
  // public history?: HistoryService;
  // public logger: DLogger;
  public config: IntermediateDendronConfig;
  // public hooks: DHookDict;

  static create(state: Required<EngineState>) {
    return new WebEngine(state);
  }
  constructor(state: Required<EngineState>) {
    const { notes, wsRoot, schemas, vaults, config } = state;
    this.notes = notes;
    this.wsRoot = wsRoot;
    this.schemas = schemas;
    this.vaults = vaults;
    this.configRoot = wsRoot;
    this.config = config;
  }
}
