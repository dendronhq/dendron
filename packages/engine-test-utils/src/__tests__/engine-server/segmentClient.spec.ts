import { DendronError, RespV2 } from "@dendronhq/common-all";
import {
  SegmentClient,
  SegmentEventProps,
  TelemetryStatus,
  tmpDir
} from "@dendronhq/common-server";
import fs, { writeFileSync } from "fs-extra";
import path from "path";
import sinon from "sinon";
import { TestEngineUtils } from "../../engine";

describe("SegmentClient", () => {
  afterEach(() => {
    sinon.restore();
  });
  test("enabled by default", (done) => {
    // this is only for the SegmentClient, telemetry as a whole respects VSCode settings by default
    TestEngineUtils.mockHomeDir();

    const instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(false);

    done();
  });

  test("enabled by command", (done) => {
    TestEngineUtils.mockHomeDir();
    SegmentClient.enable(TelemetryStatus.ENABLED_BY_COMMAND);

    const instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(false);

    done();
  });

  test("enabled by config", (done) => {
    TestEngineUtils.mockHomeDir();
    SegmentClient.enable(TelemetryStatus.ENABLED_BY_CONFIG);

    const instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(false);

    done();
  });

  test("disabled by command", (done) => {
    TestEngineUtils.mockHomeDir();
    SegmentClient.disable(TelemetryStatus.DISABLED_BY_COMMAND);

    const instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(true);

    done();
  });

  test("disabled by vscode config", (done) => {
    TestEngineUtils.mockHomeDir();
    SegmentClient.disable(TelemetryStatus.DISABLED_BY_VSCODE_CONFIG);

    const instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(true);

    done();
  });

  test("disabled by workspace config", (done) => {
    TestEngineUtils.mockHomeDir();
    SegmentClient.disable(TelemetryStatus.DISABLED_BY_WS_CONFIG);

    const instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(true);

    done();
  });

  test("still recognizes legacy disable", (done) => {
    TestEngineUtils.mockHomeDir();
    const disablePath = SegmentClient.getDisableConfigPath();
    writeFileSync(disablePath, "");

    const instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(true);

    done();
  });

  test("enable command overrides legacy disable", (done) => {
    TestEngineUtils.mockHomeDir();
    const disablePath = SegmentClient.getDisableConfigPath();
    writeFileSync(disablePath, "");

    SegmentClient.enable(TelemetryStatus.ENABLED_BY_COMMAND);
    const instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(false);

    done();
  });

  test("disabled then enabled with command", (done) => {
    TestEngineUtils.mockHomeDir();
    SegmentClient.disable(TelemetryStatus.DISABLED_BY_COMMAND);

    let instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(true);

    SegmentClient.enable(TelemetryStatus.ENABLED_BY_COMMAND);

    instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(false);

    done();
  });

  test("enabled then disabled with command", (done) => {
    TestEngineUtils.mockHomeDir();
    SegmentClient.enable(TelemetryStatus.ENABLED_BY_COMMAND);

    let instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(false);

    SegmentClient.disable(TelemetryStatus.DISABLED_BY_COMMAND);

    instance = SegmentClient.instance({ forceNew: true });
    expect(instance.hasOptedOut).toEqual(true);

    done();
  });

  describe("WHEN: Enabled by CLI command", () => {
    test("THEN: recognized as enabled", (done) => {
      TestEngineUtils.mockHomeDir();
      SegmentClient.enable(TelemetryStatus.ENABLED_BY_CLI_COMMAND);
      const instance = SegmentClient.instance({ forceNew: true });
      expect(instance.hasOptedOut).toBeFalsy();
      done();
    });
  });

  describe("WHEN: Disabled by CLI command", () => {
    test("THEN: recognized as disabled", (done) => {
      TestEngineUtils.mockHomeDir();
      SegmentClient.disable(TelemetryStatus.DISABLED_BY_CLI_COMMAND);
      const instance = SegmentClient.instance({ forceNew: true });
      expect(instance.hasOptedOut).toBeTruthy();
      done();
    });
  });

  describe("WHEN: Enabled by CLI as default", () => {
    test("THEN: recognized as enabled", (done) => {
      TestEngineUtils.mockHomeDir();
      SegmentClient.enable(TelemetryStatus.ENABLED_BY_CLI_DEFAULT);
      const instance = SegmentClient.instance({ forceNew: true });
      expect(instance.hasOptedOut).toBeFalsy();
      done();
    })
  })
});

type flushResponse = {
  successCount: number,
  nonRetryableErrorCount: number,
  retryableErrorCount: number
}

describe("GIVEN a SegmentClient", () => {
  const filepath = path.join(tmpDir().name, "test.log");

  const instance = SegmentClient.instance({
    forceNew: true,
    cachePath: filepath,
  });

  const payloadThatWillSend = {
    event: "mockWillSend",
    properties: {
      hello: "world",
    },
  };

  const payloadThatWillNotSend = {
    event: "mockFailToSend",
    properties: {
      hello: "world",
    },
  };

  const invalidPayload = "this is invalid JSON";

  function mockedTrackInternal(
    event: string,
    _context: any,
    _payload: { [key: string]: any },
    _timestamp?: Date
  ): Promise<RespV2<SegmentEventProps>> {
    return new Promise<RespV2<SegmentEventProps>>((resolve) => {
      if (event === "mockFailToSend") {
        resolve({
          error: new DendronError({
            message: "Mock - Failed to send event " + event,
          }),
        });
      } else {
        resolve({ error: null });
      }
    });
  }

  beforeEach(() => {
    sinon.stub(instance, <any>"trackInternal").callsFake(mockedTrackInternal);
  });
  
  afterEach(() => {
    sinon.restore();
  });

  describe("WHEN we track an event and we can connect to Segment backend", () => {
    beforeEach(async () => {
      await instance.track("mockWillSend", {
        hello: "world",
      },);
    });

    afterEach(() => {
      if (fs.pathExistsSync(filepath)) {
        fs.removeSync(filepath);
      }
    });

    test("THEN residual cache should be empty", async (done) => {
      expect(fs.pathExistsSync(filepath)).toBeFalsy();
      done();
    });
  });

  describe("WHEN we track an event but we cannot connect to Segment backend", () => {
    beforeEach(async () => {
      await instance.track("mockFailToSend", {
        hello: "world",
      },);
    });

    afterEach(() => {
      if (fs.pathExistsSync(filepath)) {
        fs.removeSync(filepath);
      }
    });

    test("THEN residual cache should be non-empty", async (done) => {
      const fileContents = fs.readFileSync(filepath, 'utf-8');
      expect(fileContents).toMatchSnapshot();
      done();
    });
  });

  describe("WHEN a valid payload exists and we try to flush it while we can connect to Segment backend", () => {
    let results: flushResponse;

    beforeEach(async () => {
      await instance.writeToResidualCache(filepath, payloadThatWillSend);
      results = await instance.tryFlushResidualCache();
    })

    afterEach(() => {
      if (fs.pathExistsSync(filepath)) {
        fs.removeSync(filepath);
      }
    });

    test("THEN data should be sent", (done) => {
      expect(results.successCount).toEqual(1);
      expect(results.nonRetryableErrorCount).toEqual(0);
      expect(results.retryableErrorCount).toEqual(0);
      done();
    });

    test("AND the file should be empty afterward", async (done) => {
      const fileContents = fs.readFileSync(filepath, 'utf-8');
      expect(fileContents).toEqual('');
      done();
    });
  });

  describe("WHEN a valid payload exists and we try to flush it but we cannot connect to Segment backend", () => {
    let results: flushResponse;

    beforeEach(async () => {
      await instance.writeToResidualCache(filepath, payloadThatWillNotSend);
      results = await instance.tryFlushResidualCache();
    });

    afterEach(() => {
      if (fs.pathExistsSync(filepath)) {
        fs.removeSync(filepath);
      }
    });

    test("THEN data should not be sent", async (done) => {
      expect(results.successCount).toEqual(0);
      expect(results.nonRetryableErrorCount).toEqual(0);
      expect(results.retryableErrorCount).toEqual(1);
      done();
    });

    test("AND the file should keep the payload contents (for a later retry)", async (done) => {
      const fileContents = fs.readFileSync(filepath, 'utf-8');
      expect(fileContents).toMatchSnapshot();

      done();
    });
  });

  describe("WHEN an unparsable payload exists and we try to flush it", () => {
    let results: flushResponse;

    beforeEach(async () => {
      await instance.writeToResidualCache(
        filepath,
        invalidPayload as unknown as SegmentEventProps
      );

      results = await instance.tryFlushResidualCache();
    });

    afterEach(() => {
      if (fs.pathExistsSync(filepath)) {
        fs.removeSync(filepath);
      }
    });

    test("THEN data should not be sent", async (done) => {
      expect(results.successCount).toEqual(0);
      expect(results.nonRetryableErrorCount).toEqual(1);
      expect(results.retryableErrorCount).toEqual(0);
      done();
    });

    test("AND the file should still be empty afterward (not worth retrying unparsable data)", (done) => {
      const fileContents = fs.readFileSync(filepath, 'utf-8');
      expect(fileContents).toEqual('');
      done();
    });
  });

  describe("WHEN a valid payload exists and we try to flush it, some data can get sent but some doesn't (flaky connection)", () => {
    let results: flushResponse;

    beforeEach(async () => {
      await instance.writeToResidualCache(
        filepath,
        invalidPayload as unknown as SegmentEventProps
      );

      await instance.writeToResidualCache(filepath, payloadThatWillSend);
      await instance.writeToResidualCache(filepath, payloadThatWillNotSend);

      results = await instance.tryFlushResidualCache();
    })

    afterEach(() => {
      if (fs.pathExistsSync(filepath)) {
        fs.removeSync(filepath);
      }
    });

    test("THEN some data should be sent", (done) => {
      expect(results.successCount).toEqual(1);
      expect(results.nonRetryableErrorCount).toEqual(1);
      expect(results.retryableErrorCount).toEqual(1);
      done();
    });

    test("AND the file should keep the payload contents of ONLY data that was not sent", (done) => {
      const fileContents = fs.readFileSync(filepath, 'utf-8');
      expect(fileContents).toMatchSnapshot();
      done();
    });
  });
});