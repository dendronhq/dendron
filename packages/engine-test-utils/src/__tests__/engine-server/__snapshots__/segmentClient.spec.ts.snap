// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN a SegmentClient WHEN a valid payload exists and we try to flush it but we cannot connect to Segment backend AND the file should keep the payload contents (for a later retry) 1`] = `"{\\"event\\":\\"mockFailToSend\\",\\"properties\\":{\\"hello\\":\\"world\\"}}"`;

exports[`GIVEN a SegmentClient WHEN a valid payload exists and we try to flush it, some data can get sent but some doesn't (flaky connection) AND the file should keep the payload contents of ONLY data that was not sent 1`] = `"{\\"event\\":\\"mockFailToSend\\",\\"properties\\":{\\"hello\\":\\"world\\"}}"`;

exports[`GIVEN a SegmentClient WHEN we track an event but we cannot connect to Segment backend THEN residual cache should be non-empty 1`] = `
"{\\"event\\":\\"mockFailToSend\\",\\"properties\\":{\\"hello\\":\\"world\\"}}
"
`;
